src/App.tsx

```diff
@@ -1,26 +1,341 @@
+      .catch(() => {
+        if (canEdit === true) {
+          setCanEdit(true);
+        }
+
+        handleError(errorType.deleteTask);
+        setDeletingIds([]);
+      });
+  };
+
+  const clearCompleted = () => {
+    Promise.all(
+      completedTodos.map(todo => {
+        deleteTask(todo.id);
+        setIsUpdating([]);
+      }),
+    ).then(() => {
+      focusInput();
+    });
+  };
+
+  const handleCompleted = (id: number) => {
+    setIsUpdating(current => [...current, id]);
+    const todoToUpdate = tasks.find(todo => todo.id === id);
+
+    if (!todoToUpdate) {
+      setErrorMessage(errorType.found);
+
+      return;
+    }
+
+    const updatedTodo = { ...todoToUpdate, completed: !todoToUpdate.completed };
+
+    updateCompletedTodo(id, updatedTodo)
+      .then(() => {
+        const updatedTasks = tasks.map(todo =>
+          todo.id === id ? updatedTodo : todo,
+        );
+
+        setTasks(updatedTasks);
+        setIsUpdating([]);
+      })
+      .catch(() => {
+        handleError(errorType.updateTodo);
+        setIsUpdating([]);
+      });
+  };
+
+  useEffect(() => {
+    getTodos()
+      .then(setTasks)
+      .then(() => focusInput())
+      .catch(() => handleError(errorType.load));
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, []);
+
+  useEffect(() => {
+    const timeoutId = setTimeout(() => setErrorMessage(''), 3000);
+
+    return () => clearTimeout(timeoutId);
+  }, [errorMessage]);
+
+  useEffect(() => {
+    if (taskLeft > 0) {
+      return setAllDone(false);
+    } else if (taskLeft == 0) {
+      return setAllDone(true);
+    }
+  }, [taskLeft]);
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">
-          React Todo App - Add and Delete
-        </a>
-      </p>
-
-      <p className="subtitle">Styles are already copied</p>
-    </section>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <ToDoHeader
+          handleAllCompleted={handleAllCompleted}
+          allDone={allDone}
+          handleSubmit={handleSubmit}
+          taskTitle={taskTitle}
+          setTaskTitle={setTaskTitle}
+          IsSubmitting={IsSubmitting}
+          inputRef={inputRef}
+          taskLengthForButton={tasks.length}
+        />
+        <TodoList
+          filteredTodos={filteredTodos}
+          handleCompleted={handleCompleted}
+          tempTodo={tempTodo}
+          deleteTask={deleteTask}
+          deletingIds={deletingIds}
+          onUpdate={isUpdating}
+          setNewTitle={setNewTitle}
+          newTitle={newTitle}
+          handleError={handleError}
+          setTasks={setTasks}
+          setErrorMessage={setErrorMessage}
```

Comment:
create handlers for all setState instead of passing it as props 
[[more info]](https://blog.stackademic.com/dont-pass-setstate-as-a-prop-2cc2b187d323)